pkg/ring/incremental_transfer.go

```diff
@@ -0,0 +1,391 @@
+		op = Write
+	}
+
+	var lastErr error
+
+	// Collect all the ranges that we will transfer data for
+	// and the remote ingester to send or receieve the data.
+	//
+	// When joining, the target ingester is the old end of the
+	// replica set (i.e., the last in the set) and when leaving,
+	// the target is the new end of the replica set.
+	for replica := 0; replica < rf; replica++ {
+		// The original token ranges we might want to request data
+		// for is defined by [token-(replica+1), token-replica). If ingesters
+		// have multiple tokens, there may be up to #tokens*replica number of
+		// combinations here for which our token is the replica'th successor.
+		endRanges, err := d.Predecessors(NeighborOptions{
+			Start:        token,
+			Offset:       replica,
+			Op:           op,
+			MaxHeartbeat: i.cfg.RingConfig.HeartbeatTimeout,
+		})
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		for _, endRange := range endRanges {
+			startRange, err := d.Successor(NeighborOptions{
+				Start:        endRange.StatefulToken(),
+				Offset:       -1,
+				Op:           op,
+				MaxHeartbeat: i.cfg.RingConfig.HeartbeatTimeout,
+			})
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			target, err := d.Successor(NeighborOptions{
+				Start:        startRange.StatefulToken(),
+				Offset:       rf,
+				Op:           op,
+				MaxHeartbeat: i.cfg.RingConfig.HeartbeatTimeout,
+			})
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			// No transfer necessary if we're in the replica set. If we're
+			// joining, it means we already have the data. If we're leaving,
+			// it means we should still have the data.
+			//
+			// We want to check the replica set _ignoring_ our current token,
+			// which we do by checking two ranges: [endRange, token) and
+			// (token, target].
+
+			lhsRange := RangeOptions{
+				Range: TokenRange{From: endRange.Token, To: token.Token},
+
+				ID:            i.ID,
+				Op:            op,
+				LeftInclusive: true,
+				MaxHeartbeat:  i.cfg.RingConfig.HeartbeatTimeout,
+			}
+			rhsRange := RangeOptions{
+				Range: TokenRange{From: token.Token, To: target.Token},
+
+				ID:             i.ID,
+				Op:             op,
+				RightInclusive: true,
+				MaxHeartbeat:   i.cfg.RingConfig.HeartbeatTimeout,
+			}
+
+			if d.InRange(lhsRange) || d.InRange(rhsRange) {
+				continue
+			}
+
+			addr := d.Ingesters[target.Ingester].Addr
+			ret[addr] = append(ret[addr], TokenRange{
+				From: startRange.Token,
+				To:   endRange.Token,
+			})
+		}
+	}
+
+	if lastErr != nil {
+		level.Error(util.Logger).Log(
+			"msg", fmt.Sprintf("failed to find complete transfer set for %d", token.Token),
+			"err", lastErr,
+		)
+	}
+	return ret, lastErr == nil
+}
+
+// joinIncrementalTransfer will attempt to incrementally obtain chunks from
+// neighboring lifecyclers that contain data for token ranges they will
+// no longer receive writes for.
+func (i *Lifecycler) joinIncrementalTransfer(ctx context.Context) error {
```

Comment:
This method is too long, and would really benefit from some refactoring. For example handling of individual token can be extracted into a separate method? Same goes for `joinIncrementalTransfer`.

Reply:
Sure, I'll do that.